src/harness/testcases/heartbeat_testcase.py

```diff
@@ -824,3 +824,123 @@ def test_WINNF_FT_S_HBT_18(self):
+    self.assertEqual(response['grantId'], grant_id)
+    self.assertEqual(response['response']['responseCode'], 105)
+    del request, response
+
+  @winnforum_testcase
+  def test_WINNF_FT_S_HBT_20(self):
+    """Heartbeat Request from CBSDs in Granted or Authorized state 
+    (immediately after CBSDs moves into Granted State or following
+     a Heartbeat Response) requires CBSDs to de-register. 
+
+    The response should be FAIL, code 105.
+    """
+    # Register the devices
+    devices = ['device_a', 'device_e', 'device_c']
+    device_list = []
+    for device in devices:
+      device_val = json.load(open(os.path.join('testcases', 'testdata', '{0}.json'.format(device))))
+      self._sas_admin.InjectFccId({'fccId': device_val['fccId']})
+      device_list.append(device_val)
+
+    request = {'registrationRequest': device_list}
+    response = self._sas.Registration(request)['registrationResponse']
+    # Check registration response
+    for resp in response:
+      self.assertEqual(resp['response']['responseCode'], 0)
+    cbsd_ids = [resp['cbsdId'] for resp in response]
+    del request, response
+
+    # Request grant
+    grant_list = []
+    for cbsd_id in cbsd_ids:
+      grant = json.load(open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+      grant['cbsdId'] = cbsd_id
+      grant_list.append(grant)
+    request = {'grantRequest': grant_list}
+
+    # Check grant response
+    response = self._sas.Grant(request)['grantResponse']
+    self.assertEqual(len(response), len(cbsd_ids))
+    for resp_number, resp in enumerate(response):
+      self.assertEqual(resp['cbsdId'], cbsd_ids[resp_number])
+      self.assertTrue(resp['grantId'])
+      self.assertEqual(resp['response']['responseCode'], 0)
+    grant_ids = [resp['grantId'] for resp in response]
+    del request, response
+
+    # Inject Third Device into Blacklist
+    self._sas_admin.BlacklistByFccId({'fccId': device_list[2]['fccId']})
+
+    # First Heartbeat Request
+    heartbeat_list = []
+    for cbsd_id, grant_id in zip(cbsd_ids, grant_ids):
+      heartbeat_list.append({
+          'cbsdId': cbsd_id,
+          'grantId': grant_id,
+          'operationState': 'GRANTED'
+      })
+
+    request = {'heartbeatRequest': heartbeat_list}
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    self.assertEqual(len(response), len(grant_ids))
+
+    # Check the heartbeat response
+    # First two devices are not in Blacklist must have Response Code 0
+    for index, resp in enumerate(response[:2]):
+      self.assertEqual(resp['cbsdId'], cbsd_ids[index])
+      self.assertEqual(resp['grantId'], grant_ids[index])
+      self.assertEqual(resp['response']['responseCode'], 0)
+
+    # Last Device in Blacklist must have Response Code 105
+    self.assertEqual(response[2]['cbsdId'], cbsd_ids[2])
+    self.assertEqual(response[2]['grantId'], grant_ids[2])
+    self.assertEqual(response[2]['response']['responseCode'], 105)
+
```

Comment:
I've added a check for transmitExpireTime for the non-success response as well. I missed that it is a required field in Heartbeat Response Object based on WINNF-16-S-0016 v1.0.0. Thanks for the correction!